### ISSUE TITLE
Sequential calls to realm.write() sometimes throw an objc exception

### ISSUE THREAD

**Issue by jlj** [realm/realm-cocoa#4858]
In the context of a TestFlight beta-test, I get a significant number (a few each day) of crash reports similar to this one:

````
Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Triggered by Thread:  0

Last Exception Backtrace:
0   CoreFoundation                	0x2596b916 __exceptionPreprocess + 122 (NSException.m:162)
1   libobjc.A.dylib               	0x25106e12 objc_exception_throw + 34 (objc-exception.mm:531)
2   Realm                         	0x520e68 -[RLMRealm beginWriteTransaction] + 180 (RLMRealm.mm:446)
3   RealmSwift                    	0x3eb150 _TFC10RealmSwift5Realm5writefzFzT_T_T_ + 44 (Realm.swift:153)
4   LapMonitor                    	0xe80fc _TTSf4g_n___TFC10LapMonitor16ActiveLapSessionP33_35B0DADE8A11F20A2053872E3981933117updateSessionDatafT20forDeviceSessionInfoVS_21LapMonitorSessionInfo_T_ + 972 (LapSession.swift:639)
5   LapMonitor                    	0xeafc8 _TTSf4d_g_n___TFC10LapMonitor16ActiveLapSession10lapMonitorfTCS_10LapMonitor20didUpdateSessionInfoVS_21LapMonitorSessionInfo_T_ + 360 (LapSession.swift:0)
6   LapMonitor                    	0xe13d8 _TTWC10LapMonitor16ActiveLapSessionS_24LapMonitorSessionHandlerS_FS1_10lapMonitorfTCS_10LapMonitor20didUpdateSessionInfoVS_21LapMonitorSessionInfo_T_ + 112 (LapSession.swift:0)
7   LapMonitor                    	0xd0c60 _TTSf4g_n___TFC10LapMonitor10LapMonitorP33_25576D73CF51C6E590B13A2BA772807319handleStatusMessagefVS0_P33_25576D73CF51C6E590B13A2BA772807323LapMonitorStatusMessageT_ + 564 (LapMonitor.swift:263)
8   LapMonitor                    	0xd25c0 _TTSf4d_g_d_n___TFC10LapMonitor10LapMonitor10peripheralfTCSo12CBPeripheral17didUpdateValueForCSo16CBCharacteristic5errorGSqPs5Error___T_ + 836 (LapMonitor.swift:0)
9   LapMonitor                    	0xcea44 _TToFC10LapMonitor10LapMonitor10peripheralfTCSo12CBPeripheral17didUpdateValueForCSo16CBCharacteristic5errorGSqPs5Error___T_ + 68 (LapMonitor.swift:0)
10  CoreBluetooth                 	0x2aa70cc0 -[CBPeripheral handleAttributeEvent:args:attributeSelector:delegateSelector:delegateFlag:] + 124 (CBPeripheral.m:644)
11  CoreBluetooth                 	0x2aa70d5c -[CBPeripheral handleCharacteristicEvent:characteristicSelector:delegateSelector:delegateFlag:] + 68 (CBPeripheral.m:666)
12  CoreBluetooth                 	0x2aa70e96 -[CBPeripheral handleCharacteristicValueUpdated:] + 70 (CBPeripheral.m:703)
13  CoreBluetooth                 	0x2aa769ee __29-[CBXpcConnection handleMsg:]_block_invoke + 58 (CBXpcConnection.m:227)
14  libdispatch.dylib             	0x254d981e _dispatch_call_block_and_release + 6 (init.c:757)
15  libdispatch.dylib             	0x254d980a _dispatch_client_callout + 18 (init.c:817)
16  libdispatch.dylib             	0x254e7ba4 _dispatch_main_queue_callback_4CF$VARIANT$mp + 1520 (inline_internal.h:1063)
17  CoreFoundation                	0x2592db68 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 4 (CFRunLoop.c:1612)
18  CoreFoundation                	0x2592c062 __CFRunLoopRun + 1570 (CFRunLoop.c:2718)
19  CoreFoundation                	0x2587b224 CFRunLoopRunSpecific + 516 (CFRunLoop.c:2814)
20  CoreFoundation                	0x2587b010 CFRunLoopRunInMode + 104 (CFRunLoop.c:2844)
21  GraphicsServices              	0x26e6bac4 GSEventRunModal + 156 (GSEvent.c:2245)
22  UIKit                         	0x29f4f184 UIApplicationMain + 140 (UIApplication.m:3772)
23  LapMonitor                    	0x6d208 main + 132 (LapMonitorData.swift:12)
24  libdyld.dylib                 	0x2552386e tlv_get_addr + 42 (threadLocalHelpers.s:311)
````

All Realm write transaction in the app use `realm.write` and are called from the main thread, so it seems very unlikely that a previous write transaction was left open.

I could not so far reproduce the issue in a debug environment, so any advice about how to add instrumentation to the code so that this issue can be further investigated via crash logs is welcome. 😀

Note that this looks very similar to #3831, that was supposed to be fixed in v1.0.2.

Any idea of what in the app could cause this ?

This issue is problematic because -if my understanding is correct - the objc exception thrown by `-[RLMRealm beginWriteTransaction]`  can not be catched in Swift , so  the app will inevitably crash when this occurs.

## Version of Realm and Tooling

Realm framework version: RealmSwift 2.4.x, RealmSwift 2.5.1

Realm Object Server version: n/a

Xcode version: Xcode 8.2, Xcode 8.3.1

iOS versions / devices:
- iOS 9.3.5 (13G36) -  iPhone4,1
- iOS 10.2.1 (14D27) - iPhone7,2
- iOS 10.3 (14E277) - iPhone7,2
- iOS 10.3.1 (14E304) - iPhone6,2  iPhone7,2 iPhone8,4

**Comment by istx25** [comment]
Hi @jlj. I wanted to let you know that we've received your issue and that someone will review what you've shared and follow-up soon.

**Comment by jlj** [comment]
I finally succeeded in catching this objc exception under debugger (with the exact same callstack as listed above). 

This confirms that the (default and only) realm considers to be in a write transaction:

````
2017-04-14 19:58:02.645282+0200 LapMonitor[1341:277860] *** Terminating app due to uncaught exception 'RLMException', reason: 'Cannot create asynchronous query while in a write transaction'
````

But:

 - I triple-checked that every write transaction in the app is encapsulated in a call to `realm.write {}`.
- The exception callstack does not show any sign of a recursive call to `realm.write {}`.

Do I miss something obvious here?

**Comment by TimOliver** [comment]
Hi @jlj!

Hmm, that error message would seem to indicate you're trying to set up a notification block inside a write transaction. Can you confirm if that's the case? :)

Thanks!

**Comment by jlj** [comment]
Hi @TimOliver 

No,  I have already checked that all notification blocks in the app are **not** set up inside a write transactions.

Fact which is btw confirmed by the callstack, since, as mentioned earlier in this thread, all write transaction are encapsulated by passing closures to `realm.write()` calls, and we would certainly see, if this were the case, more than one call to realm in the callstack. :)

**Comment by jlj** [comment]
@TimOliver @jpsim Are there cases where `realm.write()` could leave a write transaction uncommitted after returning?

**Comment by TimOliver** [comment]
@jlj Yes, that's possible. We've had issues in the past before where a user somehow ended up with a write transaction that failed on `commitWrite()` and so it stayed open to the point where it caused an exception when a second write transaction attempt was made. We weren't ever able to work with them to isolate why the write transaction was failing.

That being said, the exception message was different in that case (that one was failing since it said it's not possible to open 2 write transactions at the same time on the same thread), so I'm not sure if this is actually the same problem.

**Comment by jlj** [comment]
@TimOliver Thanks for your answer. I agree with you that the past issue that you mention is a different problem, as I have not seen any sign of a failed `commitWrite()` in my case.

On my side, I am pretty confident that I have identified the sequence in my app that caused the exception: under specific circumstances, two successive `realm.write()` transactions were executed sequentially in the same run loop occurence (triggered by the same Bluetooth message) and - in rare cases - the  second write transaction threw an exception in `-[RLMRealm beginWriteTransaction]` because `isInWriteTransaction` was still tested as true.

This is confirmed by the exception callstack, which always corresponds to the second write transaction.

Which means that there are cases in which  `realm.write()` succeeds and returns without throwing any error, but during some time after it returns, `isInWriteTranaction` can still be tested as true, and starting a new write transaction may crash the program.

*An interesting fact is that  the crash was reported to be much more frequent (compared to the number of application sessions) on some device types, and especially on the old and slower iPhones 4S.*

To confirm this analysis, I have released a new version of my app under TestFlight, that combine the two successive write transactions into one. No crash reported so far, but let's wait a couple of days to be sure. :)

**Comment by jlj** [comment]
I just renamed the issue with a more specific title. 😎

**Comment by jlj** [comment]
> To confirm this analysis, I have released a new version of my app under TestFlight, that combine the two successive write transactions into one. No crash reported so far, but let's wait a couple of days to be sure. :)

This is confirmed: merging the two sequential write transactions removes the crash, as shown by build 25 in this TestFlight screenshot (every crash in the 3 previous builds had a callstack identical to the one in the first message above).

![Screen capture 2017-04-19 a 17 01 52](https://cloud.githubusercontent.com/assets/1850262/25188087/29aa5d82-2525-11e7-9f47-3e176abd2eb7.png)

So I suppose that further investigations about this issue should now be on the Realm team side. :)

**Comment by jpsim** [comment]
Yeah, we can't do anything with this until we can reproduce it. It sounds impossible and there's nothing that jumps out as causing this in our code.

**Comment by misha83** [comment]
I have the same problem. When i try set property in first transaction(by .write) and second transaction add new object with the same runloop app crashes. When i merge transaction everything works fine. Realm refresh helps.

**Comment by MikeKasperlik** [comment]
We also see this in our app. The circumstances are very similar the ones @jlj described.

I was able to track this down further to the following lines in our code:

```
    func delete(object: Object) {
        autoreleasepool {
            if let realm = try? Realm(), !object.isInvalidated {
                realm.beginWrite() // Here we crash, but only some time after try realm.commitWrite() failed for a yet unknown reason
                realm.delete(object)

                do {
                    try realm.commitWrite() // This seems to fail sometimes
                } catch {
                    Log.e("Error while deleting object")
                }
            }
        }
    }
```

At first, I was curios how a realm which was opened a line before `realm.beginWrite()` could be in a transaction already. This is until I noticed, if there is something failing when calling `realm.commitWrite()` one of the NEXT calls to the `delete(object: Object)` fails with this stacktrace. 

Is a aborted transaction not canceled? Do I need to cancel it via `realm.cancelWrite()` in the catch?

**Comment by jpsim** [comment]
If `Realm.commitWrite()` throws an Error, the write transaction is automatically canceled and `Realm.isInWriteTransaction` will return `false`. We have some unit tests that validate this behavior, such as this one: https://github.com/realm/realm-cocoa/blob/v2.8.3/Realm/Tests/RealmTests.mm#L1353-L1374

**Comment by MikeKasperlik** [comment]
I did some field tests (as I can't reproduce this in the testing env). When I add a 

```
if realm.isInWriteTransaction {
    realm.cancelWrite()
}
```

in front of the `realm.beginWrite()` these errors seem to be less frequent. Currently we have like 0.5-1% of our sessions crashing with that. So far (after a day of testing) only around 0.05% of all sessions see this. 

I will do another test and add a `realm.cancelWrite()` in the catch block and see if these errors further reduce. 

If I call `Realm()` two times on the same thread - do I get the same realm instance? Or two independent? I guess first.

**Comment by jlj** [comment]
I can see this problem again, and it still looks related with very close back-to-back write transactions done in the main thread.

All realm write transactions done in the app now use the instrumented method below:

````
extension Realm {
    
    func doWriteTransaction(file : String = #file, line: Int = #line, function: String = #function, _ block: (() throws -> Void)) {
        
        let logTransactionString = "Realm.doWriteTransaction called from \(function) (\(file) line \(line))"
        NSLogDebug("%@", logTransactionString)
        
        precondition(!isInWriteTransaction, "realm.write() was called when isInWriteTransaction == true")
        
        do {
            try self.write (block)
        }
        catch let error as NSError {
            let errorString = "Realm write error in \(function) (\(file) line \(line)): \(error)"
            NSLog("%@", errorString)
            preconditionFailure(errorString)
        }
    }
}
````

Note the `precondition` call that checks that the realm doesn't consider to be in a write transaction. This precondition never triggers.

However, I still sometimes crash on the same objc exception, on the `try  self.write (block)`.

Here is the trace when this occurs (with the NSLog and the exception):

````
2017-07-24 22:46:58.362042+0200 LapMonitor[2642:913141] Realm.doWriteTransaction called from lapMonitor(_:didReceiveLapEvent:) (/Users/jean-luc/Developpement/Xcode apps/iOS apps/LapMonitor/LapMonitor/LapSession.swift line 934)
2017-07-24 22:50:04.049373+0200 LapMonitor[2642:913141] *** Terminating app due to uncaught exception 'RLMException', reason: 'Cannot create asynchronous query while in a write transaction'
````

And the callstack:

````
(lldb) thread backtrace
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x000000018e811014 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000018e8db264 libsystem_pthread.dylib`pthread_kill + 112
    frame #2: 0x000000018e7859c4 libsystem_c.dylib`abort + 140
    frame #3: 0x000000018e2511b0 libc++abi.dylib`abort_message + 132
    frame #4: 0x000000018e26ac04 libc++abi.dylib`default_terminate_handler() + 304
    frame #5: 0x000000018e278820 libobjc.A.dylib`_objc_terminate() + 124
    frame #6: 0x000000018e2675d4 libc++abi.dylib`std::__terminate(void (*)()) + 16
    frame #7: 0x000000018e266ef8 libc++abi.dylib`__cxa_throw + 136
    frame #8: 0x000000018e27866c libobjc.A.dylib`objc_exception_throw + 364
    frame #9: 0x000000010096b6d4 Realm`-[RLMRealm beginWriteTransaction] + 80
    frame #10: 0x00000001005ea5e8 RealmSwift`RealmSwift.Realm.write (() throws -> ()) throws -> () + 48
  * frame #11: 0x000000010012ac90 LapMonitor`Realm.doWriteTransaction(file="/Users/jean-luc/Developpement/Xcode apps/iOS apps/LapMonitor/LapMonitor/LapSession.swift", line=934, function="lapMonitor(_:didReceiveLapEvent:)", block=0x0000000100117ba0 LapMonitor`partial apply forwarder for LapMonitor.ActiveLapSession.(lapMonitor (LapMonitor.LapMonitor, didReceiveLapEvent : LapMonitor.LapMonitorEventInfo) -> ()).(closure #1) at LapSession.swift, self=0x000000017022c080) throws -> ()) -> () at RealmWrite.swift:21
    frame #12: 0x000000010011624c LapMonitor`ActiveLapSession.lapMonitor(lapMonitor=0x00000001701b3a20, lapEventInfo=LapMonitor.LapMonitorEventInfo @ 0x000000016fdba130, self=0x00000001700de450) -> () at LapSession.swift:942
    frame #13: 0x0000000100116e54 LapMonitor`protocol witness for LapMonitorSessionHandler.lapMonitor(LapMonitor, didReceiveLapEvent : LapMonitorEventInfo) -> () in conformance ActiveLapSession at LapSession.swift:0
    frame #14: 0x00000001000ea154 LapMonitor`LapMonitor.handleLapEventMessage(lapEventMessage=LapMonitor.LapMonitor.LapEventMessage @ 0x000000016fdba3a8, self=0x00000001701b3a20) -> () at LapMonitor.swift:309
    frame #15: 0x00000001000eed14 LapMonitor`LapMonitor.peripheral(peripheral=0x00000001702e9a80, characteristic=0x00000001702ad0e0, error=nil, self=0x00000001701b3a20) -> () at LapMonitor.swift:639
    frame #16: 0x00000001000ef014 LapMonitor`@objc LapMonitor.peripheral(CBPeripheral, didUpdateValueFor : CBCharacteristic, error : Error?) -> () at LapMonitor.swift:0
    frame #17: 0x00000001966f02e4 CoreBluetooth`-[CBPeripheral handleAttributeEvent:args:attributeSelector:delegateSelector:delegateFlag:] + 248
    frame #18: 0x00000001966f0444 CoreBluetooth`-[CBPeripheral handleCharacteristicEvent:characteristicSelector:delegateSelector:delegateFlag:] + 132
    frame #19: 0x00000001966ecd3c CoreBluetooth`-[CBPeripheral handleMsg:args:] + 320
    frame #20: 0x00000001966e7de8 CoreBluetooth`-[CBCentralManager handleMsg:args:] + 204
    frame #21: 0x00000001966fbe18 CoreBluetooth`-[CBManager xpcConnectionDidReceiveMsg:args:] + 116
    frame #22: 0x00000001966f46a0 CoreBluetooth`__30-[CBXpcConnection _handleMsg:]_block_invoke + 60
    frame #23: 0x00000001028a1a50 libdispatch.dylib`_dispatch_call_block_and_release + 24
    frame #24: 0x00000001028a1a10 libdispatch.dylib`_dispatch_client_callout + 16
    frame #25: 0x00000001028af2e8 libdispatch.dylib`_dispatch_queue_serial_drain + 1140
    frame #26: 0x00000001028a5634 libdispatch.dylib`_dispatch_queue_invoke + 852
    frame #27: 0x00000001028a69d4 libdispatch.dylib`_dispatch_main_queue_callback_4CF + 784
    frame #28: 0x000000018f7c50c8 CoreFoundation`__CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
    frame #29: 0x000000018f7c2ce4 CoreFoundation`__CFRunLoopRun + 1572
    frame #30: 0x000000018f6f2da4 CoreFoundation`CFRunLoopRunSpecific + 424
    frame #31: 0x000000019115c074 GraphicsServices`GSEventRunModal + 100
    frame #32: 0x00000001959ad058 UIKit`UIApplicationMain + 208
    frame #33: 0x00000001000a9fbc LapMonitor`main at AppDelegate.swift:12
    frame #34: 0x000000018e70159c libdyld.dylib`start + 4
````

Realm framework version: RealmSwift 2.8.3

**How can we progress on this?**

**Comment by stuartro** [comment]
I'm running into the exact same problem, with RealmSwift 2.8.3. Actions triggered from UI cause sequential (non-overlapping) realm.write { ... } blocks to be executed, and without fail the second realm.write { ... } block will fail with exception. 

Interestingly enough, if I call realm.refresh() before starting the second realm.write { ... }, then the crash goes away.

**Comment by stuartro** [comment]
This seems to be related to notifications (and, yes, I'm dead sure that I'm not adding a notification inside of a write transaction), as it only occurs if I have subscriptions (via RxRealm) to property changes on an object being modified.

**Comment by evilutioner** [comment]
We faced with the same problem, 100% reproduced in our project. Steps to reproduce:
1. Create by RXSwift two observables and concat them, perform on MainScheduler
2. Both observable should be as deferred and should write/delete items in realm
3. Before this create notification subscription to the fields that modified in step 2
4. Subscription should check numbers of items in realm as reloadData() does.

Expected result: SIGABRT

**Comment by stuartro** [comment]
I had the same problem, but found a fairly simple solution that is working very well for me. Basically, check  if `isInWriteTransaction` and, if so, first call `realm.refresh()`:

```public func write(_ block: ((Realm) throws -> Swift.Void)) throws {
   let realm = self.realm()
   if realm.isInWriteTransaction {
      try! block(realm)
   } else {
      realm.doWriteTransaction {
         try! block(realm)
      }
   }
}
```

```
extension Realm {
   public func doWriteTransaction(file: String = #file, line: Int = #line, function: String = #function, _ block: (() throws -> Void)) {
      // Work around the crash that occurs when two write transactions are called in the same runloop.
      refresh()
      precondition(!isInWriteTransaction, "realm.write() was called when isInWriteTransaction == true")
      do {
         try self.write(block)
      } catch let error as NSError {
         let errorString = "Realm write error in \(function) (\(file) line \(line)): \(error)"
         ... handleErrorAnyWayYouPrefer(error, ...)
      }
   }
}
```

**Comment by stuartro** [comment]
Realm, RxSwift, etc. versions I'm using are as follows:

```
github "realm/realm-cocoa"               == 3.7.5 
github "RxSwiftCommunity/RxRealm"        == 0.7.5
github "ReactiveX/RxSwift"               == 4.2.0
github "RxSwiftCommunity/RxSwiftExt"     == 3.3.0
```

**Comment by tgoyne** [comment]
It sounds like #4818 is the remaining problem here so closing this as a duplicate of that.
